LibGit2Sharp/Core/ManagedHttpSmartSubtransport.cs

```diff
@@ -0,0 +1,278 @@
+
+                    if (readLen == 0)
+                    {
+                        break;
+                    }
+
+                    postBuffer.Write(buffer, 0, readLen);
+                    length -= readLen;
+                    writeTotal += readLen;
+                }
+
+                if (writeTotal < length)
+                {
+                    throw new EndOfStreamException("Could not write buffer (short read)");
+                }
+
+                return 0;
+            }
+
+            private bool CertificateValidationProxy(object sender, X509Certificate cert, X509Chain chain, SslPolicyErrors errors)
+            {
+                try
+                {
+                    int ret = SmartTransport.CertificateCheck(new CertificateX509(cert), (errors == SslPolicyErrors.None), EndpointUrl.Host);
+                    Ensure.ZeroResult(ret);
+
+                    return true;
+                }
+                catch(Exception e)
+                {
+                    SetError(e);
+                    return false;
+                }
+            }
+
+            private string GetUserAgent()
+            {
+                string userAgent = GlobalSettings.GetUserAgent();
+
+                if (string.IsNullOrEmpty(userAgent))
+                {
+                    userAgent = "LibGit2Sharp " + GlobalSettings.Version.InformationalVersion;
+                }
+
+                return userAgent;
+            }
+
+            private HttpRequestMessage CreateRequest(Uri endpointUrl, bool isPost, string contentType)
+            {
+                var verb = isPost ? new HttpMethod("POST") : new HttpMethod("GET");
+                var request = new HttpRequestMessage(verb, endpointUrl);
+                request.Headers.Add("User-Agent", String.Format("git/2.0 ({0})", GetUserAgent()));
+                request.Headers.Remove("Expect");
+
+                return request;
+            }
+
+            private HttpResponseMessage GetResponseWithRedirects()
+            {
+                ICredentials credentials = null;
+                var url = EndpointUrl;
+                int retries;
+
+                for (retries = 0; ; retries++)
+                {
+                    var httpClientHandler = CreateClientHandler();
+                    httpClientHandler.Credentials = credentials;
+
+                    using (var httpClient = new HttpClient(httpClientHandler))
+                    {
+                        var request = CreateRequest(url, IsPost, ContentType);
+
+                        if (retries > MAX_REDIRECTS)
+                        {
+                            throw new Exception("too many redirects or authentication replays");
+                        }
+
+                        if (IsPost && postBuffer.Length > 0)
+                        {
+                            var bufferDup = new MemoryStream(postBuffer.GetBuffer());
+                            bufferDup.Seek(0, SeekOrigin.Begin);
```

Comment:
This causes trailing NUL bytes in the request as the complete buffer is forwarded without the accompanying length restriction. Also, no need to seek a newly created stream to position 0 as that's the default behavior anyway. So you could use this instead:

```suggestion
                            var bufferDup = new MemoryStream(postBuffer.GetBuffer(), 0, (int) postBuffer.Length);
```

When I scanned over the code, it just looked a bit strange to me. But with this simple change, I cannot reproduce the dreaded "early EOF" error anymore :)

Reply:
Stupendous find! Awesome.

@ethomson let's get this and #1772 in and complete this puppy 🐶.